flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ExecutionJobVertexTaskInputsOutputsDescriptorBuilder.java

```diff
@@ -0,0 +1,119 @@
+			getMaxPossibleNumbersOfInputChannelsPerInstance(executionJobVertex);
+		final Map<IntermediateDataSetID, Integer> maxPossibleNumbersOfResultSubpartitionsPerInstance =
+			getMaxPossibleNumbersOfResultSubpartitionsPerInstance(executionJobVertex, vertices);
+
+		return TaskInputsOutputsDescriptor.from(
+			maxPossibleNumbersOfInputChannelsPerInstance,
+			maxPossibleNumbersOfResultSubpartitionsPerInstance);
+	}
+
+	private static Map<IntermediateDataSetID, Integer> getMaxPossibleNumbersOfInputChannelsPerInstance(
+			final ExecutionJobVertex executionJobVertex) {
+
+		Map<IntermediateDataSetID, Integer> maxPossibleNumbers = new HashMap<>();
+		final List<JobEdge> inputEdges = executionJobVertex.getJobVertex().getInputs();
+		for (int i = 0; i < inputEdges.size(); i++) {
+			final JobEdge inputEdge =  inputEdges.get(i);
+			final IntermediateResult consumedResult = executionJobVertex.getInputs().get(i);
+
+			// the inputs order should match in JobGraph and ExecutionGraph
+			checkState(consumedResult.getId().equals(inputEdge.getSourceId()));
+
+			final int maxPossibleNumber = getMaxPossibleNumberOfEdgesToTarget(
+				executionJobVertex.getParallelism(),
+				consumedResult.getNumberOfAssignedPartitions(),
+				inputEdge.getDistributionPattern());
+			maxPossibleNumbers.put(consumedResult.getId(), maxPossibleNumber);
+		}
+
+		return maxPossibleNumbers;
+	}
+
+	private static Map<IntermediateDataSetID, Integer> getMaxPossibleNumbersOfResultSubpartitionsPerInstance(
+			final ExecutionJobVertex executionJobVertex,
+			final Map<JobVertexID, ExecutionJobVertex> vertices) {
+
+		Map<IntermediateDataSetID, Integer> maxPossibleNumbers = new HashMap<>();
+		final List<IntermediateDataSet> producedDataSets = executionJobVertex.getJobVertex().getProducedDataSets();
+		for (int i = 0; i < producedDataSets.size(); i++) {
+			final IntermediateDataSet producedDataSet =  producedDataSets.get(i);
+
+			checkState(
+				producedDataSet.getConsumers().size() == 1,
+				new IllegalStateException("Currently in execution, a result should have exactly one consumer job vertex."));
+
+			final JobEdge outputEdge = producedDataSet.getConsumers().get(0);
+			final ExecutionJobVertex consumerJobVertex = vertices.get(outputEdge.getTarget().getID());
+
+			final int maxPossibleNumber = getMaxPossibleNumberOfEdgesToTarget(
+				executionJobVertex.getParallelism(),
+				consumerJobVertex.getParallelism(),
+				outputEdge.getDistributionPattern());
+			maxPossibleNumbers.put(producedDataSet.getId(), maxPossibleNumber);
+		}
+
+		return maxPossibleNumbers;
+	}
+
+	private static int getMaxPossibleNumberOfEdgesToTarget(
+			final int thisParallelism,
+			final int targetParallelism,
+			final DistributionPattern distributionPattern) {
+
+		switch (distributionPattern) {
+			case ALL_TO_ALL:
+				return targetParallelism;
+			case POINTWISE:
+				return (int) Math.ceil((double) targetParallelism / thisParallelism);
```

Comment:
Yes. Currently we have to do this due to dynamic slot sharing. A shared slot is possible to host a combination of tasks requiring max possible shuffle memory, and we cannot foretell this will happen in which slot due to dynamic slot sharing.
Fortunately the shuffle memory waste is not critical here since it is POINTWISE. Network buffers required for POINTWISE shuffle is actually much fewer than in the ALL_TO_ALL case. The shuffle memory depends on edge count, which is M * N in ALL_TO_ALL case, but just max{M, N} in POINTWISE case.